### PR TITLE
fix: WebSocket connection-manager concurrency + proxy-aware rate limiting

### PR DESCRIPTION
**Pull request by williaby**
## Summary

**Tier 2** of the architecture review. **Stacked on #61** — base is the PR #61 branch, so this diff shows only the tier-2 changes. Rebase/retarget to `main` once #61 merges.

### C3 — WebSocket `ConnectionManager` concurrency safety
The previous manager had two real races when a broadcast overlapped a connect/disconnect on the same batch:

- **"Set changed size during iteration"** — `broadcast()` iterated the live connection set while `await`-ing each `send_json`; a concurrent `disconnect` mutating that set raised mid-iteration. Now it iterates a **snapshot** (`list(...)`).
- **Key resurrection** — cleanup used `defaultdict` access (`self._connections[key].discard(...)`), which *recreated* a batch key that a concurrent `disconnect` had just emptied and removed, leaking entries. Switched to a plain `dict` with **non-resurrecting cleanup** (`.get`/`.pop`) in both `broadcast` and `disconnect`.

Rationale for not adding an `asyncio.Lock`: under CPython's single-threaded event loop, `connect`/`disconnect` mutations are synchronous (atomic w.r.t. other coroutines); the only interleaving point is the awaited sends in `broadcast`, which the snapshot + non-resurrecting cleanup fully cover. A lock would force `disconnect` to become async (rippling into the router and tests) for no additional correctness here.

### H6 — proxy-aware rate limiting
Behind Cloudflare, `request.client.host` is the *proxy* IP, so per-IP rate limiting effectively keyed every request on one IP. `RateLimitMiddleware` can now resolve the client IP from a configured header:

- New settings `rate_limit_trust_proxy` (**default off**) and `rate_limit_client_ip_header` (default `CF-Connecting-IP`).
- **Default off is deliberate**: the header is only trusted when behind a proxy that overwrites it; otherwise a client could spoof it to evade limits. When off, behavior is unchanged (`request.client.host`).
- Comma-separated forwarding headers (e.g. `X-Forwarded-For`) use the first (originating) entry.

## Verification

| Gate | Result |
|------|--------|
| Tests | **456 passed**, 1 skipped (+8 new) |
| Coverage | **92.16%** (gate: 80%) |
| ruff check + format | clean |
| basedpyright | 0 errors |

New tests: broadcast tolerates concurrent disconnect; broadcast doesn't resurrect a removed batch; proxy IP resolution (trusted/untrusted/missing-header/forwarded-list/no-client) + an integration test proving distinct `CF-Connecting-IP` values are limited independently.

## Test plan
- [x] `uv run pytest` — 456 passed, 1 skipped
- [x] `uv run ruff check .` / `ruff format --check`
- [x] `uv run basedpyright src/` — 0 errors
- [x] Coverage ≥ 80% (actual 92.16%)

https://claude.ai/code/session_01PA6dtgMhfzSe22VVtqBfxE

---
_Generated by [Claude Code](https://claude.ai/code/session_01PA6dtgMhfzSe22VVtqBfxE)_

### CODE REVIEW

**Comment by williaby** [comment]
### PR Review

Core engineering (C3 concurrency + H6 proxy resolution) is correct and well-tested. An independent adversarial pass confirmed the ConnectionManager concurrency fix holds under the single-threaded event-loop model, and the 8 new tests cover both regressions plus all proxy edge cases. The items below should be addressed before merge.

#### Critical
- **Merge state: DIRTY.** This branch has conflicts with `main` and is stacked on the unmerged #61. Merge #61 first, then rebase/retarget this PR to `main` and resolve conflicts. Do not merge before #61 lands.

#### Important
- **Rate-limit key cardinality (defense-in-depth):** [`middleware/security.py` `_get_client_ip`](https://github.com/ByronWilliamsCPA/rag-processor/blob/009ad7bde7b17a5c2b8830c1680491e2c173a6c0/src/rag_processor/middleware/security.py#L169-L213). When `trust_proxy_headers=True`, the rate-limit key becomes the high-cardinality real client IP. `max_tracked_ips` is enforced only inside `_cleanup_stale_entries`, which early-returns until `cleanup_interval` (default 300s) elapses, while `dispatch` inserts keys with no inline cap. Between cleanup cycles an attacker controlling many source addresses (e.g. an IPv6 /64 faithfully forwarded by the proxy, no spoofing needed) can inflate the tracking table past `max_tracked_ips`. The off-by-default mode is unaffected. Suggested fix: enforce `max_tracked_ips` inline on insert in `dispatch` (evict-on-write), independent of the time-gated cleanup; optionally validate the forwarded value with `ipaddress.ip_address(...)` before keying on it. The time-gated cleanup itself is pre-existing; H6 is what makes it reachable.
- **PR description understates scope.** The body states the diff shows "only the tier-2 changes," but the base is `main` with #61 unmerged, so the diff also carries tier-1 work (`core/redis.py` pool factory, the `ProjectBaseError` exception handler, ingest persist+enqueue, `queue/jobs.py` rollback/async wrappers/worker-failure handling, the `cloudflare.py` JWKS refactor, `time_utils` extraction). Make the "do not merge before #61" ordering explicit.
- **CHANGELOG.md not updated.** This is a `fix:` PR; the OpenSSF baseline requires a CHANGELOG entry for fix/feat/perf/breaking changes.

#### Resolved / Informational
- Both Copilot comments (blank-leading-IP fallback; `SecurityConfig` docstring fields) are already addressed by the latest commit.
- CodeRabbit auto-review was skipped because the base/target was not the default branch at creation; retargeting to `main` should re-enable it.

Copilot review present; see the Reviewers section for results.

Generated with [Claude Code](https://claude.ai/code)

**Comment by williaby** [comment]
### PR Fix Summary

Resolved the conflict and addressed every review finding. Branch rebased onto `main` (now that #61 is merged) and force-pushed; the diff is now strictly tier-2.

**Conflict / scope**
- Rebased `claude/tier-2-concurrency` `--onto origin/main`, dropping #61's tier-1 commits that `main` already has as a squash. `mergeable` is now `MERGEABLE`. The diff is now only the C3 + H6 changes (6 files), matching the PR description. The full CI suite (Code Quality, CodeQL, Compatibility Matrix, E2E, Newman, SonarCloud) is running for the first time; it had silently no-op'd while the PR was conflicting.

**Code fixes**
- `fix(security)` (`24d67ee`): bound the rate-limit tracking table on insert. When `trust_proxy_headers` is enabled the key becomes the real client IP, so a flood of distinct IPs could grow `self.requests` without limit between the time-gated periodic cleanups (`max_tracked_ips` was enforced only inside `_cleanup_stale_entries`). Now enforced on the request fast path via `O(1)` insertion-order eviction (`_evict_oldest_if_over_cap`), independent of the periodic activity-LRU trim. Non-IP values in the trusted proxy header are rejected (`ipaddress` validation) so they cannot be injected as tracking keys; they fall back to the peer address. Added regression tests (50-distinct-IP flood stays bounded; non-IP entry falls back).
- `docs` (`143d1ae`): documented the `ConnectionManager` lock-free concurrency invariant in the class docstring; added CHANGELOG `[Unreleased]` entries for the tier-2 work.

**Verification (local)**
| Gate | Result |
|------|--------|
| ruff check / format | clean |
| basedpyright | 0 errors |
| pytest | 462 passed, 1 skipped |
| coverage | 92.19% |

**Already resolved / no action**
- Copilot's two comments (blank-leading-IP fallback; `SecurityConfig` docstring) were already addressed by an earlier commit and carried through the rebase.
- Two prior-review Suggested items (`_run_pipeline` COMPLETED guard; router async-wrapper consistency) were tier-1 code that is now on `main`, so they are out of scope for this PR.

**Environment note (not a PR defect):** `uv sync --all-extras` fails locally because the `dev` extra bundles `atheris`, which needs the LLVM/libFuzzer toolchain to build. This blocks one-command setup of type-check/test tooling on machines without clang. Worth pulling `atheris` into a separate optional extra (e.g. `fuzz`) so `dev` installs cleanly.

Generated with [Claude Code](https://claude.ai/code)

File: src/rag_processor/middleware/security.py [hunk 1]
```diff
@@ -147,15 +150,57 @@ def __init__(
...
+            forwarded = request.headers.get(self.client_ip_header)
+            if forwarded:
+                # X-Forwarded-For-style headers may list multiple hops; the
+                # first is the originating client.
+                return forwarded.split(",")[0].strip()
+            logger.warning(
+                "trust_proxy_headers enabled but %s header missing; "
+                "falling back to direct peer address",
+                self.client_ip_header,
+            )
```
[the remark this answers is not in the document]
Comment (williaby) [reply src/rag_processor/middleware/security.py]: Fixed in 6da55fc (carried through the rebase): the blank/non-IP leading entry now falls back to the peer address, and the SecurityConfig proxy fields are documented. Hardened further in 24d67ee with ipaddress validation and an O(1) tracking-table cap.

File: src/rag_processor/middleware/security.py [hunk 2]
```diff
@@ -508,6 +548,8 @@ class SecurityConfig:
     allowed_origins: list[str] = field(default_factory=list)
     allowed_hosts: list[str] = field(default_factory=list)
     rate_limit_rpm: int = 60
+    trust_proxy_headers: bool = False
+    client_ip_header: str = "CF-Connecting-IP"
```
[the remark this answers is not in the document]
Comment (williaby) [reply src/rag_processor/middleware/security.py:630]: Fixed in 6da55fc (carried through the rebase): the blank/non-IP leading entry now falls back to the peer address, and the SecurityConfig proxy fields are documented. Hardened further in 24d67ee with ipaddress validation and an O(1) tracking-table cap.
